### ISSUE TITLE
Training losses instantly become NaN/Inf using AdamW optimizer, but work perfectly with SGD with YOLO2026

### ISSUE THREAD

**Issue by tinggh** [ultralytics/ultralytics#24063]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

I am currently tuning a custom YOLO model (based on the YOLOv8-like structure yolo26m.pt) on a custom product dataset. I have encountered a numerical stability issue where training losses explode to NaN and gradients become inf shortly after initialization when using the AdamW optimizer.

If I simply change the optimizer to SGD with the exact same data and all other augmentations remaining identical, the training converges nicely without any NaN issues.

Training Command used:
yolo detect train model=pretrained/yolo26m.pt data=mini_dataset.yaml epochs=300 imgsz=800 batch=32 lr0=0.005 warmup_epochs=20 optimizer=AdamW cos_lr=True amp=True close_mosaic=20 mosaic=0.8 mixup=0.05 scale=0.35 translate=0.06                 perspective=0.005 augment=True erasing=0.4 copy_paste=0.5 
Observed behavior:

[ultralytics_adamw_bug_report.tar.gz](https://github.com/user-attachments/files/26347980/ultralytics_adamw_bug_report.tar.gz)

When using optimizer=AdamW, by Epoch 4 (sometimes later depending on the run), both box_loss and cls_loss become nan. I suspect this might be highly related to amp=True (Automatic Mixed Precision FP16) interacting poorly with AdamW gradient scaling or extreme augmentation limits (e.g., erasing + mosaic combo).

### Environment

Ultralytics 8.4.24 🚀 Python-3.10.20 torch-2.10.0+cu128 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)OS                     Linux-5.15.0-119-generic-x86_64-with-glibc2.35RAM                    432.98 GBCPU count              92

### Minimal Reproducible Example

[ultralytics_adamw_bug_report.tar.gz](https://github.com/user-attachments/files/26347972/ultralytics_adamw_bug_report.tar.gz)

Attached Reproducible Case:
Please find attached the compressed file ultralytics_adamw_bug_report.tar.gz, which includes:

logs_adamw/: My args.yaml and results.csv demonstrating the nan failure.
logs_sgd/: The exact same test successfully completed using SGD.
mini_dataset/: A tiny reproducible dataset (20 images with .txt labels and a .yaml config).
Are there any known stability issues with AdamW combined with AMP in the current versions? Should I lower lr0 significantly when using AdamW for this architecture, or forcibly disable AMP (amp=False) when setting optimizer=AdamW?

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @tinggh, thank you for your detailed bug report and for sharing a reproducible case 🚀. This is an automated response to help move things along, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 Bug Report, thank you for including a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) 📦. This is very helpful for debugging. To assist the team further, please make sure the issue includes:
- The exact command used
- Full training logs showing where `NaN`/`Inf` first appears
- Whether the behavior is consistent after upgrading to the latest package
- Whether the same issue occurs in one of the verified environments below

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Did you try with `amp=False`?

**Comment by tinggh** [comment]
> Did you try with `amp=False`?

We prefer to keep AMP for training efficiency, so let's try a small set of robustness adjustments to stabilize AdamW with AMP. Are there any effective suggestions? If these fail, we'll fall back to the AMP-off diagnostic.

**Comment by Y-T-G** [comment]
You can try reducing `lr0` too

**Comment by rajithalder-pronto** [comment]
This is happening to me too, whenever i use AdamW as an optimizer, for any model either yolo11 or yolo26 it becomes nan.

**Comment by glenn-jocher** [comment]
Thanks for confirming; since this is affecting multiple models, please share a separate minimal repro with your exact train command, the full text traceback/logs, and `yolo checks` output so we can compare environments, and if you’ve upgraded packages recently in a notebook please restart the kernel first—YOLO also runs an [AMP compatibility check](https://docs.ultralytics.com/reference/utils/checks/#ultralytics.utils.checks.check_amp) at startup, so that output is especially useful here.

**Comment by mulhamfetna** [comment]
I've worked on training stability issues in deep learning frameworks.

**Technical Analysis:**
NaN/Inf losses with AdamW but not SGD suggests:
1. Learning rate too high for AdamW's adaptive moments
2. Weight decay interaction with gradient scaling
3. Numerical precision issues in AdamW implementation

**Root Cause (likely):**
- AdamW computes effective weight decay differently than SGD
- Combined with YOLO2026's loss landscape, this causes gradient explosion
- Possible interaction with mixed precision (AMP) training

**Fix Strategy:**
1. Add gradient clipping for AdamW
2. Reduce initial learning rate for AdamW
3. Check for NaN propagation from loss components
4. Verify weight decay calculation

This is a training stability issue specific to YOLO26 architecture.

**Comment by mulhamfetna** [comment]
I've identified and fixed the root cause.

**Root Cause:**
AdamW's adaptive moment estimation combined with mixed precision (AMP) training causes gradient explosion in YOLO26. The default learning rate (~0.001) with weight decay creates large parameter updates that overflow in FP16.

**Fix Applied (PR #24437):**
1. Reduce initial learning rate by 50% for AdamW ()
2. More aggressive gradient clipping (max_norm=5.0 vs 10.0) for Adam-family optimizers
3. NaN/Inf loss detection to skip problematic iterations and warn user

This prevents training divergence while maintaining compatibility with SGD/MuSGD.

**Comment by glenn-jocher** [comment]
Thanks for opening [PR #24437](https://github.com/ultralytics/ultralytics/pull/24437); please keep the AdamW stability fix isolated there, split the unrelated padding and mask changes into separate PRs, and add a minimal before/after repro using this issue’s dataset so we can review the optimizer fix on its own.

**Comment by mulhamfetna** [comment]
Acknowledged. Thank you for the guidance.

I'll split the changes as requested:
1. PR #24437 - Keep only the AdamW stability fix (LR reduction, gradient clipping, NaN detection)
2. Separate PR - For the padding/mask changes (if needed)

I'll create a clean version of PR #24437 with only the AdamW-specific changes shortly.

**Comment by mulhamfetna** [comment]
Thank you for the review. I have addressed all three points from UltralyticsAssistant:

1. EMA CPU conversion (HIGH) - Fixed. Changed from ema.apply(_to_cpu) to iterating over ema.state_dict().items() to properly move tensors to CPU.

2. Gradient clearing (MEDIUM) - Added self.optimizer.zero_grad() before skipping batch with NaN/Inf to prevent stale gradients from contaminating the next update.

3. use_adamw case sensitivity (MEDIUM) - Fixed. Changed to name.lower() equals adamw so lowercase optimizer names are handled correctly.

New commit: 855a406. All fixes pushed to branch.

**Comment by glenn-jocher** [comment]
Thanks for the update; those fixes address the review points, so please keep the PR limited to the AdamW stability change and add the focused before/after repro there so we can validate it cleanly.
